puzzle-utils/words/ladder.py

```diff
@@ -0,0 +1,148 @@
+from .search import search
+
+def ladder(word):
+    """
+    Given a word, find all "adjacent" words in the dictionary
+    i.e. all words with edit distance of 1 from the given word
+
+    Params:
+    word (required) - The base word
+
+    Alternatives:
+    find_ladders        - Find all ladders between two words
+    find_ladders_pretty - Print all unique ladders between two words nicely
+    """
+    found_words = set()
+    for i in range(len(word)):
+        pattern = list(word)
+        pattern[i] = '.'
+        search_results = search("^" + "".join(pattern) + "$")
+        for result in search_results:
+            if result != word and result not in found_words:
+                found_words.add(result)
+    return found_words
+
+def edit_distance(left_word, right_word):
+    """
+    This assumes the lengths of the words are the same.
+    The result is the edit distance of the two words (number of
+    substitutions required to tranform left_word into right_word).
+    """
+    if len(left_word) != len(right_word):
+        raise ValueError("Word ladder words must be same length")
+
+    distance = 0;
+    for i in range(len(left_word)):
+        if left_word[i] != right_word[i]:
+            distance += 1
+    return distance
+
+def follow_ladder_path(word_path, right_word, cached_searches, depth, exact_depth, max_depth, use_heuristic):
+    """
+    Helper to recursively find a word ladder path
+    """
+    # We're too deep, this will take too long
+    if depth > max_depth:
+        return None
+
+    # We don't care about this, we want a specific depth
+    if exact_depth is not None and depth > exact_depth:
+        return None
+
+    # Find words closer to destination (avoiding cycles)
+    current_word = word_path[-1]
+    max_edit_distance = edit_distance(current_word, right_word) if use_heuristic else len(current_word)
+    adjacent_words = []
+    if current_word in list(cached_searches.keys()):
+        adjacent_words = cached_searches[current_word]
+    else:
+        adjacent_words = list(ladder(current_word))
+        cached_searches[current_word] = adjacent_words
+    next_words = list(filter(lambda w: w not in word_path and edit_distance(w, right_word) <= max_edit_distance, adjacent_words))
+
+    # No ladder exists this way
+    if len(next_words) == 0:
+        return None
+
+    # We found a ladder! Append if it's the right depth
+    if right_word in next_words:
```

Comment:
I believe we should handle the path found case by checking if `current_word == right_word` in the beginning of the function as opposed to checking if `right_word in next_words` here. I see that you're handling the base case of the recursion here. In general, you want the base case to be the "0th case", but yours here is like the "1st case", if that makes sense. Aside from the conceptual clarity, this approach also gracefully handles the case if `left_word` is equal to `right_word` which I believe your code wouldn't.